src/p_keyframe.c

```diff
@@ -0,0 +1,587 @@
+}
+
+static void UnArchiveWorld(void)
+{
+    int i;
+    sector_t *sec;
+    line_t *li;
+    side_t *si;
+
+    // do sectors
+    for (i = 0, sec = sectors; i < numsectors; i++, sec++)
+    {
+        read32(&sec->floorheight,
+               &sec->ceilingheight,
+               &sec->floor_xoffs,
+               &sec->floor_yoffs,
+               &sec->ceiling_xoffs,
+               &sec->ceiling_yoffs,
+               (int32_t *)&sec->floor_rotation,
+               (int32_t *)&sec->ceiling_rotation);
+
+        read16(&sec->floorpic,
+               &sec->ceilingpic,
+               &sec->lightlevel,
+               &sec->special,
+               &sec->tag);
+
+        // Woof!
+        sec->soundtarget = readp();
+        sec->thinglist = readp();
+        sec->floordata = readp();
+        sec->ceilingdata = readp();
+        sec->lightingdata = readp();
+    }
+
+    // do lines
+    for (i = 0, li = lines; i < numlines; i++, li++)
+    {
+        read16((int16_t *)&li->flags,
+               &li->special,
+               &li->tag);
+        
+        read32((int32_t *)&li->angle,
+               &li->frontmusic,
+               &li->backmusic);
+
+        // Woof!
+        li->frontsector = readp();
+        li->backsector = readp();
+
+        for (int j = 0; j < 2; j++)
+        {
+            if (li->sidenum[j] != NO_INDEX)
+            {
+                si = &sides[li->sidenum[j]];
+              
+                read16(&si->toptexture,
+                       &si->bottomtexture,
+                       &si->midtexture);
+                
+                read32(&si->textureoffset,
+                       &si->rowoffset);
+                si->oldtextureoffset = si->textureoffset;
+                si->oldrowoffset = si->rowoffset;
+            }
+        }
+    }
+}
+
+static void ArchivePlayState(keyframe_t *keyframe)
+{
+    writex(&thinkercap, sizeof(thinkercap), 1);
+    writex(thinkerclasscap, sizeof(thinker_t), NUMTHCLASS);
+    keyframe->thinkers = M_CopyArena(thinkers_arena);
+    writep(headsecnode);
+    keyframe->msecnodes = M_CopyArena(msecnodes_arena);
+    writex(&activeceilings, sizeof(*activeceilings), 1);
+    keyframe->activeceilings = M_CopyArena(activeceilings_arena);
+    writex(&activeplats, sizeof(*activeplats), 1);
+    keyframe->activeplats = M_CopyArena(activeplats_arena);
+
+    writex(blocklinks, blocklinks_size, 1);
+
+    write32(floatok,
+            felldown,
+            tmfloorz,
+            tmceilingz);
+    writep(ceilingline);
+    writep(floorline);
+    writep(linetarget);
+    writep(sector_list);
+    writex(tmbbox, sizeof(tmbbox), 1);
+    writep(blockline);
+    write32(hangsolid);
+
+    write32(numspechit);
+    writex(spechit, sizeof(*spechit), numspechit);
+    
+    write32(opentop,
+            openbottom,
+            openrange,
+            lowfloor);
+    writex(&trace, sizeof(trace), 1);
+    write32(num_intercepts);
+    writex(intercepts, sizeof(*intercepts), num_intercepts);
```

Comment:
The idea is to save and restore all global variables and data structures to create perfect "save states", similar to those in emulators and virtual machines. I added everything that I found in the `p_*` files, but I might have missed something or included something unnecessary.

I fixed most of the issues, the "Rewind" key should work now. I'm interested in feedback!

TODO:
1. Add support for demos and archive demo-compatible keyframes.
2. Use this system for multiplayer rollback (in other PR).

Reply:
No, I think we should keep the old save game format. This is only for rewind/multiplayer. Old saves work as before, by the way.

Reply:
What has become of the "saving will be one memcpy()" approach that you announced earlier?